make the timebar display correctly when using START values

barbeque-squared opened the pull request:
fixes #610

I moved all the timebar-drawing logic to `UScreenSingView` and made it use less variables. It should also properly respect the START value now.

Some edge cases and other things that someone should check on a current USDX before merging:
- if someone Ctrl^Left's a lot during a song with a high START value, the progress bar starts going 'negative'. I consider this more as a bug of being able to Ctrl^Left there in the first place, but it might be a regression compared to current behaviour.
- the progress bar will overflow into the time left (or whatever is in the bottom right) if the song doesn't properly end for some reason. See #600. This is a regression compared to current drawing behaviour.
- running a Medley doesn't draw anything (no progress, no note blocks) but did that ever work in the first place? IIRC only the progress bar but not the note blocks (can someone check?) and if so, not sure how useful/critical it is for this PR.

barbeque-squared commented:
(don't merge this yet, I noticed an issue with it on a song that used to work, will investigate why)

barbeque-squared commented:
and now it uses `GAP` again. should be good now.

barbeque-squared commented:
> running a Medley doesn't draw anything (no progress, no note blocks) but did that ever work in the first place? IIRC only the progress bar but not the note blocks (can someone check?) and if so, not sure how useful/critical it is for this PR.

have verified this.
before this PR: only progress bar (no note blocks)
after this PR: nothing at all

I'll fix the progress bar for medleys. The note blocks should become their own separate issue, as these have never worked anyway.

barbeque-squared commented:
@bohning Duet != Medley
Go to any song (there might be a length requirement or something, but most normal songs should work), press M -> Medley -> Sing!
![2023-03-10-185820_721x381_scrot](https://user-images.githubusercontent.com/5775429/224389639-3155d14d-60b0-4424-adbe-255896372206.png)
![2023-03-10-185830_698x360_scrot](https://user-images.githubusercontent.com/5775429/224389652-6449f591-c4f0-40d7-88f1-00427b29eca5.png)
That's the usecase my comment refers to.

bohning commented:
Ah, my bad, I misread that. I will delete my comments in order not to distract from the issue, so you can delete your answer as well.

barbeque-squared commented:
Rebased the entire branch (because otherwise I got FFmpeg version mismatch build errors).

Then I made it draw the bar also during medleys again. And because it's all in a single function now, it _also_ actually draws the sentence boxes:
![screenshot0001](https://user-images.githubusercontent.com/5775429/224494585-d35fe5ca-b492-4d23-897c-bf5db5e7eeb9.jpg)
Compared to the screenshot, I changed it so that during medleys, it will just skip drawing a box for any sentence that _starts_ earlier than the medley starts, thus preventing the overflow to the left in the screenshot.

The only other issue I could find is that at the beginning of a medley (it does this on every medley), the white progress bar goes 'negative' a little bit. I think this has to do with where it seeks to not 100% lining up with medleystart - fadeintime?

NB: the screenshot was made for an auto-generated medley on a ~40 minute song that uses #START and #END tags to play only a part of it, so it's a fairly complex test case (L-Train - Fall of an Empire - Scene 09/Ch6 : Kingdoms Divided).

I'll leave this open for a week or so, for people to test with one of the build artifacts. Test especially with medleys, because I don't ever use them. Otherwise this is going in, because I do have an actual need for normal songs being able to use START/END.

basisbit commented:
In the future, please always use merge commits instead of rebase, in order to keep the original commit timeline.
